kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/bytearray/ByteArrayRecordWriterProvider.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.connect.s3.format.bytearray;
+
+import io.confluent.connect.s3.S3SinkConnectorConfig;
+import io.confluent.connect.s3.storage.S3OutputStream;
+import io.confluent.connect.s3.storage.S3Storage;
+import io.confluent.connect.storage.format.RecordWriter;
+import io.confluent.connect.storage.format.RecordWriterProvider;
+import org.apache.kafka.connect.converters.ByteArrayConverter;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+public class ByteArrayRecordWriterProvider implements RecordWriterProvider<S3SinkConnectorConfig> {
+
+  private static final Logger log = LoggerFactory.getLogger(ByteArrayRecordWriterProvider.class);
+  private static final String EXTENSION = ".out";
+  private static final String LINE_SEPARATOR = System.lineSeparator();
+  private static final byte[] LINE_SEPARATOR_BYTES = LINE_SEPARATOR.getBytes();
+  private final S3Storage storage;
+  private final ByteArrayConverter converter;
+
+  ByteArrayRecordWriterProvider(S3Storage storage, ByteArrayConverter converter) {
+    this.storage = storage;
+    this.converter = converter;
+  }
+
+  @Override
+  public String getExtension() {
+    return EXTENSION;
+  }
+
+  @Override
+  public RecordWriter getRecordWriter(final S3SinkConnectorConfig conf, final String filename) {
+    return new RecordWriter() {
+      final S3OutputStream s3out = storage.create(filename, true);
+
+      @Override
+      public void write(SinkRecord record) {
+        log.trace("Sink record: {}", record);
+        try {
+          byte[] bytes = converter.fromConnectData(
+              record.topic(), record.valueSchema(), record.value());
```

Comment:
My thought here is to pass the `record.value()` blindly into the converter without verifying that it's a byte array since `fromConnectData` contains checks around type and will emit a DataException we catch below.